### DISCUSSION TITLE
How to self host with MX record on subdomain and use reverse proxy

### DISCUSSION THREAD

**Discussion by whisperdancer** [simple-login/app#1272] [Self-hosting question]
My domain hosts its mail with a third party mail server, zoho.com. As such my MX records for the domain point to zoho mail servers. I need to keep this functionality and ensure I don't break it setting up self host for simplelogin. Also, I have other services on my network and I use LISO Swag container as a reverse proxy to route the traffic correctly and all services get SSL. It works quite well, but I have been unable to get simplelogin to work as self host.

You readme.md file states: - a domain that you can config the DNS. It could be a sub-domain.

**MX Record**
As I don't want to break my existing mail for the domain which is working great hosted at zoho, for this to work would I need to add an MX record for a subdomain. So rather than use mydomain, I would use sub.mydomain.com? Would I then do this?

Create a **MX record** that points `sub.mydomain.com` to `app.mydomain.com.` with `priority 10`?

To verify if the DNS works, the following command

```
dig @1.1.1.1 sub.mdomain.com mx
should return:

sub.mydomain.com.	3600	IN	MX	10 app.mydomain.com.
```

I use cloudflare for DNS management for my domain and I'm unsure how to add the dkim record for sub.domain.com?

Super important! Would doing this NOT break my domain mail hosted at zoho? 

**Reverse Proxy**

I use LISO Swag container as my reverse proxy. It also handles the cert generation automatically. I setup Swag so that my domain DNS uses wildcard for subdomains. I also have a cname record for this. Swag uses proxy conf files to determine how to route the traffic properly. Out of the box they include sample configs for many popular services, but simplelogin is not included. As such, what would be the proprer nginx config to use for a reverse proxy setup? Perhaps consider working with the LISO team to provide them a sample to include with Swag "out of the box". This would be quite helpful.

I did a "best effort" and manually created simplelogin.subdomain.conf and pasted in this. I restarted swag to load in the newly added proxy.conf file

```
server {
    listen 443 ssl;
    listen [::]:443 ssl;

    server_name app.*;

    include /config/nginx/ssl.conf;

    client_max_body_size 0;

    # enable for ldap auth, fill in ldap details in ldap.conf
    #include /config/nginx/ldap.conf;

    # enable for Authelia
    #include /config/nginx/authelia-server.conf;

    location / {
        # enable for Authelia
        #include /config/nginx/authelia-location.conf;

        include /config/nginx/proxy.conf;
        include /config/nginx/resolver.conf;
        set $upstream_app 192.168.xxx.xx;
        set $upstream_port 7777;
        set $upstream_proto http;
        proxy_pass $upstream_proto://$upstream_app:$upstream_port;

    }
```

I also setup nginx on the simplelogin server per your instructions. However, when I visit https://app.mydomain.com, I get 501 bad gateway. Connection refused errors in the log

**Comment by nguyenkims** [comment]
> Super important! Would doing this NOT break my domain mail hosted at zoho?

As long as you don't modify the DNS records for your root domain, this shouldn't affect your setup at Zoho.

About Nginx, I don't think it supports wildcard `*` so I'd suggest to put the exact subdomain here instead.

The DKIM for a subdomain is set up the same way as for the root domain.

**Comment by whisperdancer** [comment]
Actually, nginx on swag does support wildcard. I use it this same way to proxy multiple other services. Is there instructions on how to self host simplelogin using a reverse proxy documented somewhere?

Thank you.

**Reply by nguyenkims** [reply]
At the end of our self host readme, you can see a simple Nginx config file.
https://github.com/simple-login/app/blob/af5f31fe5f41a0ebe7079a599ef4503b76bc5cff/README.md#L502

**Comment by whisperdancer** [comment]
Yes, I saw that and assumed that is the nginx file for nginx hosted on the same server as simpleloign. But, what if I run singplelogin behind a reverse proxy. I use swag, but it's really just encapsulates nginx under the hood. What would be the nginx file for the reverse proxy?

Thank you.

**Reply by nguyenkims** [reply]
Unfortunately we aren't familiar with swag or with running nginx on another host. SimpleLogin.io also has nginx running on the same host actually.

**Comment by whisperdancer** [comment]
The reverse proxy doesn't replace nginx running on the same server as simplelogin. It's just another instance of nginx which functions as a reverse proxy and sits on my internet facing server with open port 443. I.E my nextlcoud server runs nginx locally and swag is able to proxy the traffic to it using a proxy config file. As I can't open port 443 directly to simplelogin since that port is on my internet facing server which runs my reverse proxy, I need a way to proxy the traffic from the nginx reverse proxy to simplelogin. This requires an nginx config file to properly handle routing the traffic.

Do you have a discord channel where this can be discussed and pull in the community? I can also reference the LISO team to possibly help figure this out.

Thanks

**Reply by nguyenkims** [reply]
We don't have a discord channel and use this forum for discussing actually.

**Comment by DANgerous25** [comment]
I use HAProxy in a similar setup to you. My nginx config is incredibly simple and is below for your reference. The way I got this working was to have HAProxy forward https traffic (and convert http to https) to my web server, then I use port forwarding on my firewall to redirect SMTP 25 inbound to my web server. That way the server can receive mail (because I don’t think HAProxy works will in this scenario, at least for me):

```
$ cat /etc/nginx/sites-enabled/simplelogin
server {
    listen 443 ssl;
    listen [::]:443 ssl;

    include snippets/ca-signed.conf;

    server_name  mysub.mydomain.com;

    location / {
        proxy_pass http://localhost:7777;
    }
}
```

And my snippets file:

```
$ cat /etc/nginx/snippets/ca-signed.conf 
ssl_certificate /etc/letsencrypt/live/mydomain.com/fullchain.pem;
ssl_certificate_key /etc/letsencrypt/live/mydomain.com/privkey.pem;

ssl_protocols TLSv1.2;
ssl_prefer_server_ciphers on;
ssl_ciphers ECDHE-RSA-AES256-GCM-SHA512:DHE-RSA-AES256-GCM-SHA512:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-SHA384;
ssl_session_timeout 10m;
ssl_session_cache shared:SSL:10m;
ssl_session_tickets off;
ssl_stapling on;
ssl_stapling_verify on;
resolver 8.8.8.8 8.8.4.4 valid=300s;
resolver_timeout 5s;
add_header X-Frame-Options DENY;
add_header X-Content-Type-Options nosniff;
add_header X-XSS-Protection "1; mode=block";
```

Hope this helps you.

**Reply by whisperdancer** [reply]
My reverse proxy handles the certs so it has access to the fullchain,pem and privacy.pem. I don't have access to these files. The proxy handles the ssl. I just add a config file to inform nginx where to proxy the traffic.

**Reply by DANgerous25** [reply]
My reverse proxy does as well actually, so I guess I don’t need that snippet in this particular config. In any case, is your issue solved now or what problems are you facing?

**Reply by whisperdancer** [reply]
No, I continue to get 502 nginx error. Can't navigate to the webpage for simplelogin.

**Comment by whisperdancer** [comment]
Finally got this working. Had to remove the 127.0.0.1 from sl-app. Not needed when using reverse proxy

**Reply by DANgerous25** [reply]
Strange, I didn’t have to do that on mine. Glad you got it working.

**Reply by whisperdancer** [reply]
But even though I can now navigate to the website, I'm having other issues. Tried to create first account and get page with server error. Checked log and lots of errors. The postgres db logs show db errors. I have no idea why that is happening? And no response so far to my issue, other than some random scam account. See #1321
